layout/cycle_remover.go

```diff
@@ -56,14 +54,31 @@ func reverseEdge(g Graph, e [2]uint64) {
 }
 
 func (s SimpleCycleRemover) RemoveCycles(g Graph) {
-	for cycle := getCycle(g); len(cycle) > 0; cycle = getCycle(g) {
+	neighbors := make(map[uint64][]uint64)
+	for e := range g.Edges {
+		neighbors[e[0]] = append(neighbors[e[0]], e[1])
+	}
+	for cycle := getCycle(g.Roots(), neighbors); len(cycle) > 0; cycle = getCycle(g.Roots(), neighbors) {
 		// pick edge randomly
 		i := rand.Intn(len(cycle) - 1)
 		e := [2]uint64{cycle[i], cycle[i+1]}
 
 		reverseEdge(g, e)
 		s.Reversed[e] = true
+
+		neighbors[e[0]] = deleteValue(neighbors[e[0]], e[1])
+		neighbors[e[1]] = append(neighbors[e[1]], e[0])
+	}
+}
+
+func deleteValue(slice []uint64, value uint64) []uint64 {
+	for i, n := range slice {
+		if n == value {
+			slice[i] = slice[len(slice)-1]
+			return slice[:len(slice)-1]
+		}
 	}
+	return slice
 }
```

Comment:
very generic function. can use generics + move out to different file at least (or even package)

renaming that this delets only first one

```suggestion
func deleteFirstValue[T comparable](slice []T, value T) []T {
	for i, n := range slice {
		if n == value {
			slice[i] = slice[len(slice)-1]
			return slice[:len(slice)-1]
		}
	}
	return slice
}
```

Reply:
The project is set up to use go 1.17 and it seems generics were introduced in 1.18.
I don't know what version of go a library should use: I guess we want to use the lowest possible whereas an app would use the most recent one.

Reply:
it is ok to bump it to 1.23. no reason to keep old version